### PR TITLE
SonarQube / Jacoco

### PR DESCRIPTION
**Pull request by LeeTranMN**
Add SonarQube and Jacoco to the project. Clean up Maven output. Add uncovered code to make sure Sonar is picking up on it.

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -6,10 +6,25 @@ steps:
...
+    image: maven:3-jdk-11
+    commands:
+      - mvn -ntp -B clean
+      - mvn -ntp -B compile
   - name: test
     image: maven:3-jdk-11
     commands:
-      - mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V
-      - mvn test -B
+      - mvn -ntp install -DskipTests=true -Dmaven.javadoc.skip=true -B -V
```
Comment (chrisgoeden) [inline .drone.yml]: Why perform an install (skip the tests) and then test?  Why not perform the install with the tests and remove the mvn test line below?

Comment (chrisgoeden) [reply .drone.yml]: See Maven's Lifecycles Reference: http://maven.apache.org/ref/3.6.3/maven-core/lifecycles.html

Basically there are phases that must be performed before other phases.  In this case test is always performed before install.

Comment (LeeTranMN) [reply .drone.yml]: Dunno, that's just the way that the drone tutorial did it.

**Review by joseph-dasuki — APPROVED** [review]
_Nice_

File: .drone.yml (hunk 1 above)
Comment (chrisgoeden) [reply .drone.yml]: Looking more at the maven life cycles, it doesn't make sense to package an artifact without running unit test first.  Builds should be fast in order to provide constant feedback.  You want your build to fail as soon as something is wrong.  In this case unit tests.  We should just run the tests.  In the future we should have integration test but that should be another step.

Comment (chrisgoeden) [reply .drone.yml]: Let's remove the mvn install line and keep the mvn test line.

File: .drone.yml [hunk 2]
```diff
@@ -6,10 +6,35 @@ steps:
   - name: greeting
     image: alpine
     commands:
-      - echo hello
-      - echo world
+      - echo This step is to ensure that the pipeline itself is functioning properly.
+  - name: build
+    image: maven:3-jdk-11
+    commands:
+      - mvn -ntp -B clean
```
Comment (chrisgoeden) [inline .drone.yml]: These should be one command...mvn clean build -ntp -B -V

File: .drone.yml [hunk 3]
```diff
@@ -6,10 +6,35 @@ steps:
   - name: greeting
     image: alpine
     commands:
-      - echo hello
-      - echo world
+      - echo This step is to ensure that the pipeline itself is functioning properly.
+  - name: build
+    image: maven:3-jdk-11
```
Comment (chrisgoeden) [inline .drone.yml]: This version is too generic.  There are many versions of maven and the maintainers of this image are always upgrading it to the latest.  We should pic a specific version of maven and use that locally as well with the build machine.

This line should be:  maven:3.6.3-jdk-11

File: .drone.yml [hunk 4]
```diff
@@ -6,10 +6,35 @@ steps:
...
+    commands:
+      - mvn -ntp -B clean
+      - mvn -ntp -B compile
   - name: test
     image: maven:3-jdk-11
```
Comment (chrisgoeden) [inline .drone.yml:15]: See comments above.

**Review by chrisgoeden — CHANGES_REQUESTED** [review]
Just a couple of changes in the drone file.  Otherwise it looks good.

File: .drone.yml [hunk 5]
```diff
@@ -6,10 +6,35 @@ steps:
...
     commands:
-      - mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V
-      - mvn test -B
+      - mvn -ntp install -DskipTests=true -Dmaven.javadoc.skip=true -B -V
+      - mvn -ntp test -B
+#     Since Jacoco is running our coverage reports for us, remove Test files before sending to sonar
+      - rm -R src/test
+  - name: code-analysis
+    image: aosapps/drone-sonar-plugin
```
Comment (chrisgoeden) [inline .drone.yml]: We should use a tagged version of the image.  aosapps/drone-sonar-plugin:1.0

**Review by joseph-dasuki — APPROVED** [review]

**Review by chrisgoeden — APPROVED** [review]
Good to go!

**Pull request merged**